source/PluginHelper.cpp

```diff
@@ -0,0 +1,198 @@
+#else
+		FILE *out = fopen(location.c_str(), "wb");
+#endif
+		// Set the url that gets downloaded
+		curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
+		// Follow redirects
+		curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1l);
+		// How long we will wait
+		curl_easy_setopt(curl, CURLOPT_CA_CACHE_TIMEOUT, 604800L);
+		// What is the maximum filesize in bytes.
+		curl_easy_setopt(curl, CURLOPT_MAXFILESIZE, MAX_SIZE);
+		// Set the write function and the output file used in the write function
+		curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteData);
+		curl_easy_setopt(curl, CURLOPT_WRITEDATA, out);
+
+		CURLcode res = curl_easy_perform(curl);
+		if(res != CURLE_OK)
+			fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
+
+		curl_easy_cleanup(curl);
+		fclose(out);
+		return res == CURLE_OK;
+	}
+
+
+
+	// Copy an entry from one archive to the other
+	void CopyData(struct archive *ar, struct archive *aw)
+	{
+		int retVal;
+		const void *buff;
+		size_t size;
+		int64_t offset;
+
+		while(true)
+		{
+			retVal = archive_read_data_block(ar, &buff, &size, &offset);
+			if(retVal == ARCHIVE_EOF)
+				return;
+			if(retVal != ARCHIVE_OK)
+				return;
+			if(archive_write_data_block(aw, buff, size, offset) != ARCHIVE_OK)
+				return;
+		}
+	}
+
+
+
+	bool ExtractZIP(string filename, string destination, string expectedName)
+	{
+		int size = 0;
+
+		int flags = ARCHIVE_EXTRACT_TIME;
+		flags |= ARCHIVE_EXTRACT_PERM;
+		flags |= ARCHIVE_EXTRACT_ACL;
+		flags |= ARCHIVE_EXTRACT_FFLAGS;
+
+		// Create the handles for reading/writing
+		archive *read = archive_read_new();
+		archive *ext = archive_write_disk_new();
+		archive_write_disk_set_options(ext, flags);
+		archive_read_support_format_all(read);
+		if(!filename.empty() && strcmp(filename.c_str(), "-") == 0)
+			filename = "";
+		if(archive_read_open_filename(read, filename.c_str(), 10240))
+			return false;
+
+		// Check if this plugin has the right head folder name
+		archive_entry *entry;
+		archive_read_next_header(read, &entry);
+		string firstEntry = archive_entry_pathname(entry);
+		bool fitsExpected = firstEntry == (expectedName);
+		archive_read_data_skip(read);
+
+		// Check if this plugin has a head folder, if not create one in the destination
+		archive_read_next_header(read, &entry);
+		string secondEntry = archive_entry_pathname(entry);
+		bool hasHeadFolder = secondEntry.find(firstEntry) != std::string::npos;
+		if(!hasHeadFolder)
+#if defined(_WIN32)
+			_wmkdir(Utf8::ToUTF16(destination + expectedName).c_str());
+#else
+			mkdir((destination + expectedName).c_str(), 0777);
+#endif
+
+		// Close the archive so we can start again from the beginning
+		archive_read_close(read);
+		archive_read_free(read);
+
+		// Read another time, this time for writing.
+		read = archive_read_new();
+		archive_read_support_format_all(read);
+		archive_read_open_filename(read, filename.c_str(), 10240);
+
+		string dest_file;
+		int retVal;
```

Comment:
These two can be defined inside the while

Reply:
done